### DISCUSSION TITLE
Run common task in github action or alternative of before_script in github workflow ci

### DISCUSSION THREAD

**Discussion by uditgaurav** [community/community#27136] [Actions]
<p>Hello all,</p>
<p>I’m trying to shift my CI workflow from GitLab to Github So just having some queries:</p>
<ul>
<li>In Gitlab CI, if there are common steps running for each job we use to place it in<br>
Like:</li>
</ul>
<pre><code class="lang-auto">gitlab-ci.yml

  before_script:
    - export GOPATH=$HOME/go
    - export PATH=$PATH:/usr/local/go/bin:$GOPATH/bin
    - pwd
    - echo "Starting Tests"
</code></pre>
<p><code>before_script</code> automatically runs before the execution of every step in a job.<br>
So, Is there a way to achieve this using Github Actions?</p>
<p>Example:</p>
<pre><code class="lang-auto">jobs:
  tests:
    runs-on: ubuntu-latest
    steps:
      - name: Running for app1
        run: |
          echo "common command 1"
          echo "common command 2"
          echo "Hello world"
          
      - name: Running for app2
        run: |
          echo "common command 1"
          echo "common command 2"
          echo "Hello all"
</code></pre>

**Comment by uditgaurav** [comment]
<p>Can we place all common commands in a job that is like <code>before_Script</code>  or something like that?</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/uditgaurav">@uditgaurav</a>,<br>
Currently, GitHub Actions does not support the feature likes <strong>before_script</strong> in Gitlab CI.</p>
<blockquote>
<p><strong>Can we place all common commands in a job that is like  before_Script  or something like that?</strong></p>
</blockquote>
<p>I think this is not work. if you are using GitHub-hosted runners in to run the job in your workflow, <a href="https://help.github.com/en/actions/reference/virtual-environments-for-github-hosted-runners#about-github-hosted-runners" rel="nofollow noopener">each job in the workflow executes in a fresh instance of the virtual machine</a>. So, the system configurations (such as environment variables) you set in a job generally are not applicable to other jobs.<br>
And GitHub Actions also does not support the feature likes  <a href="https://docs.microsoft.com/en-us/azure/devops/pipelines/process/templates?view=azure-devops" rel="nofollow noopener"> <strong>Template</strong> </a> (such as job templates and step templates) for  <strong>YAML Pipelines</strong>  on  <strong>Azure Pipelines</strong> .</p>
<p>If your projects really need the similar features, I recommend you directly report a feature request <a href="https://support.github.com/contact/feedback?contact%5Bcategory%5D=actions" rel="nofollow noopener">here</a>. That will allow you to directly interact with the appropriate engineering team, and make it more convenient for the engineering team to collect and categorize your suggestions.</p>

**Comment by alwell-kevin** [comment]
<p>You can use <em>composite run steps actions</em> . These are actions that are solely defined in YAML (<a href="https://docs.github.com/en/actions/creating-actions/creating-a-composite-run-steps-action#creating-an-action-metadata-file" rel="noopener nofollow ugc">documentation</a>).</p>
<p><strong>Workflow</strong></p>
<pre><code class="lang-auto">[...]

jobs:
  job:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - uses: ./.github/workflows/composite-action

[...]
</code></pre>
<p><strong>Composite run steps action</strong><br>
<code>.github/workflows/composite-action/action.yml</code> (same repository as the workflow)</p>
<pre><code class="lang-auto">name: "My composite action"
description: "Checks out the repository and does something"
runs:
  using: "composite"
  steps: 
  - run: |
      echo "Executing action"
    shell: bash
</code></pre>
